### PR TITLE
[LM-9637] Chatroom sharing & deep linking

### PR DESCRIPTION
**Pull request by jaixchaudhary**
## Description

**Features**
- Chatroom sharing and deep linking

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Hotfix (fixes a breaking changes)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Other, please describe:

### Author Checklist

- [x] Assignee added
- [x] Reviewer added
- [x] Description added
- [ ] Unit test cases added

### Reviewer Checklist

- [x] PR title starts with Jira ticket number prefix
- [ ] If hotfix, Add Hotfix as prefix in PR title
- [x] Label added
- [x] Commit messages follows standard format (added, changed, removed, code commented and chore)
- [x] Description about the Feature/ Bugs added

### CODE REVIEW

File: ChatSX/Models/CommonFunctions/index.tsx [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export interface shareChatroomRequest {
+  chatroomId: number;
+  domain: string;
+}
```
Comment (ishaan1607) [inline ChatSX/Models/CommonFunctions/index.tsx:4]: rename `shareChatroomRequest` to `ShareChatroomRequest`

File: ChatSX/Models/CommonFunctions/index.tsx
Comment (ishaan1607) [inline ChatSX/Models/CommonFunctions/index.tsx:1]: why a separate package is created, we can add a `model` package in `commonFunctions` only.

File: ChatSX/commonFuctions/index.tsx [hunk 2]
```diff
@@ -501,3 +502,12 @@ export const formatDate = (date: any, time: any) => {
...
 };
+
+// this method generates URL for share
+export const shareChatroomURL = ({
+  chatroomId,
+  domain,
+}: shareChatroomRequest) => {
+  let URL = `${domain}/chatroom?chatroom_id=${chatroomId}`;
+  return URL;
+};
```
Comment (ishaan1607) [inline ChatSX/commonFuctions/index.tsx:513]: create a new file for sharing, names `ShareUtils`

File: ChatSX/components/ParseDeepLink/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import {myClient} from '../../..';
+import {PATH_REGEX, QUERY_REGEX, VALID_URI_REGEX} from '../../constants/Regex';
+import {initAPI} from '../../store/actions/homefeed';
+
+interface DeepLinkRequest {
+  uri: string;
+  userName: string;
+  uuid: string;
+  isGuest: boolean;
+}
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:10]: create a separate file for this interface

File: ChatSX/components/ParseDeepLink/index.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+interface DeepLinkResponse {
+  success: boolean;
+  errorMessage?: string;
+}
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:15]: create a separate file for this interface

File: ChatSX/components/ParseDeepLink/index.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    const path = matches ? matches[1] : null;
+
+    if (path === '/chatroom') {
+      let regexToExtractParams: RegExp = QUERY_REGEX;
+      let params: Record<string, string> = {};
+      let match: RegExpExecArray | null;
+
+      while ((match = regexToExtractParams.exec(uri))) {
+        params[match[1]] = match[2];
+      }
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:43]: what is the requirement for this particular loop?

File: ChatSX/components/ParseDeepLink/index.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+interface DeepLinkResponse {
+  success: boolean;
+  errorMessage?: string;
+}
+
+// this function is to check URI is valid or not
+function isValidURI(uri: string): boolean {
+  const regex = VALID_URI_REGEX;
+  return regex.test(uri);
+}
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:21]: why this necessary, as this regex check all types of links like `ftp` etc, if we want to add this check, then add check for `https://` links

File: ChatSX/components/ParseDeepLink/index.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+// this function is to parse deep link url
+async function parseDeepLink(
+  request: DeepLinkRequest,
+  responseCallback?: (response: DeepLinkResponse) => void,
+) {
+  const uri = request.uri;
+
+  if (isValidURI(uri)) {
+    const pathRegex = PATH_REGEX;
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:31]: this is unnecessary variable `PATH_REGEX` can directly be used for further operations

File: ChatSX/components/ParseDeepLink/index.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+        let initiateUserRequest = {
+          userName: request?.userName,
+          uuid: request?.uuid,
+          isGuest: false,
+        };
+        const success = await myClient.initiateUser(initiateUserRequest);
+
+        if (success) {
+          // TODO -- Redirect to internalRoute
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:58]: why TODO is added? we should add function to redirect to route?

File: ChatSX/components/ParseDeepLink/index.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  uuid: '123456',
+  isGuest: false,
+};
+
+// Example usage to call parseDeepLink() method
+export function callParseDeepLink() {
+  parseDeepLink(exampleRequest, response => {
+    // Parsed response
+  });
+}
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:101]: remove this.

File: ChatSX/screens/ChatRoom/index.tsx [hunk 10]
```diff
@@ -2349,6 +2376,9 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
                       } else if (val?.id === 28) {
                         await unblockMember();
                         setModalVisible(false);
+                      } else if (val?.id === 3) {
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:2379]: Use enums for id check, don't use directly `3`

File: ChatSX/screens/ChatRoom/index.tsx [hunk 11]
```diff
@@ -1982,6 +1983,32 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
...
+          // shared
+        }
+      } else if (result.action === Share.dismissedAction) {
+        // dismissed
+      }
+    } catch (error: any) {
+      Alert.alert(error.message);
+    }
+  };
+
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:2011]: move this code to some utils, this is not a view related code.

File: ChatSX/screens/ChatRoom/index.tsx [hunk 12]
```diff
@@ -1982,6 +1983,32 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
...
+        message: URL,
+      });
+      if (result.action === Share.sharedAction) {
+        if (result.activityType) {
+          // shared with activity type of result.activityType
+        } else {
+          // shared
+        }
+      } else if (result.action === Share.dismissedAction) {
+        // dismissed
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:2005]: why there is no code?

File: ChatSX/screens/ChatRoom/index.tsx [hunk 13]
```diff
@@ -1982,6 +1983,32 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
...
+      let URL = shareChatroomURL(shareChatroomRequest);
+
+      const result = await Share.share({
+        message: URL,
+      });
+      if (result.action === Share.sharedAction) {
+        if (result.activityType) {
+          // shared with activity type of result.activityType
+        } else {
+          // shared
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:2002]: why there is no code?

File: ChatSX/screens/ChatRoom/index.tsx [hunk 14]
```diff
@@ -1982,6 +1983,32 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
...
+        domain: 'https://dummyurl.com',
+      };
+      let URL = shareChatroomURL(shareChatroomRequest);
+
+      const result = await Share.share({
+        message: URL,
+      });
+      if (result.action === Share.sharedAction) {
+        if (result.activityType) {
+          // shared with activity type of result.activityType
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:2000]: why there is no code?

File: ChatSX/screens/ChatRoom/index.tsx [hunk 15]
```diff
@@ -1982,6 +1983,32 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
...
   };
 
+  // this function helps to share chatroom url
+  const onShare = async () => {
+    try {
+      let shareChatroomRequest = {
+        chatroomId: chatroomID,
+        domain: 'https://dummyurl.com',
+      };
+      let URL = shareChatroomURL(shareChatroomRequest);
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:1993]: rename `URL` to `shareUrl`

File: ChatSX/screens/ChatRoom/index.tsx [hunk 16]
```diff
@@ -1982,6 +1983,32 @@ const ChatRoom = ({navigation, route}: ChatRoom) => {
     }
   };
 
+  // this function helps to share chatroom url
+  const onShare = async () => {
+    try {
+      let shareChatroomRequest = {
+        chatroomId: chatroomID,
+        domain: 'https://dummyurl.com',
+      };
```
Comment (ishaan1607) [inline ChatSX/screens/ChatRoom/index.tsx:1992]: remove this code, or add todo with `domain`

File: ChatSX/store/actions/chatroom.ts [hunk 17]
```diff
@@ -131,7 +131,7 @@ export const getChatroom = (payload: any) => async (dispatch: Dispatch) => {
     return await dispatch({
       type: GET_CHATROOM_SUCCESS,
       [CALL_API]: {
-        func: myClient?.getChatroom(payload),
+        func: myClient?.getChatroomV2(payload),
```
Comment (ishaan1607) [inline ChatSX/store/actions/chatroom.ts:134]: why?

**Review by ishaan1607 — CHANGES_REQUESTED** [review]

**Comment by ishaan1607** [comment]
redirection flow is missing please add that also @jaixchaudhary

File: ChatSX/screens/ChatRoom/index.tsx (hunk 14 above)
Comment (jaixchaudhary) [reply ChatSX/screens/ChatRoom/index.tsx]: Removed this code, as these checks are used to update the status of share activity by showing toast/alert/console

File: ChatSX/screens/ChatRoom/index.tsx (hunk 12 above)
Comment (jaixchaudhary) [reply ChatSX/screens/ChatRoom/index.tsx]: Removed this code, as these checks are used to update the status of share activity by showing toast/alert/console

File: ChatSX/screens/ChatRoom/index.tsx (hunk 13 above)
Comment (jaixchaudhary) [reply ChatSX/screens/ChatRoom/index.tsx]: Removed this code, as these checks are used to update the status of share activity by showing toast/alert/console

File: ChatSX/store/actions/chatroom.ts (hunk 17 above)
Comment (jaixchaudhary) [reply ChatSX/store/actions/chatroom.ts]: updated by getChatroomActions now

File: ChatSX/screens/ChatRoom/index.tsx (hunk 11 above)
Comment (jaixchaudhary) [reply ChatSX/screens/ChatRoom/index.tsx]: Done

File: ChatSX/Models/CommonFunctions/index.tsx
Comment (jaixchaudhary) [reply ChatSX/Models/CommonFunctions/index.tsx:1]: `Models` folder is segregated, so it would be a good practice to create models in `Models` folder only

Comment (ishaan1607) [reply ChatSX/Models/CommonFunctions/index.tsx:1]: agreedm we should make a `models` folder separate, but it should be inside that specific package. For example: there will be `models` inside `commonFunctions` which will be for all the models created/related to common functions, and same for `parseDeepLink`

File: ChatSX/Models/ParseDeepLink/index.tsx [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+  userName: string;
+  uuid: string;
+  isGuest: boolean;
+}
+
+export interface DeepLinkResponse {
+  success: boolean;
+  errorMessage?: string;
+  data?: Record<string, string>;
+}
```
Comment (ishaan1607) [inline ChatSX/Models/ParseDeepLink/index.tsx:12]: create a separate file for this

File: ChatSX/constants/Numbers.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+export const TWO = 2;
+export const THREE = 3;
+export const FOUR = 4;
+export const SIX = 6;
+export const EIGHT = 8;
+export const NINE = 9;
+export const FIFTEEN = 15;
+export const TWENTY_ONE = 21;
+export const TWENTY_SEVEN = 27;
+export const TWENTY_EIGHT = 28;
```
Comment (ishaan1607) [inline ChatSX/constants/Numbers.ts:10]: what are these enum? why do we enums for counting, enums are used to specific what value it is representing, for example `id==3` represent menu it for `Share`, so enum name should `ITEM_SHARE=3` and not `THREE=3`.

Comment (ishaan1607) [inline ChatSX/constants/Numbers.ts:10]: plus, these are not enums, they are just constants, use proper enums in js/ts. Follow this: https://www.w3schools.com/typescript/typescript_enums.php

File: ChatSX/components/ParseDeepLink/index.ts [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+          userName: request?.userName,
+          uuid: request?.uuid,
+          isGuest: false,
+        };
+        const initiateUserResponse = await myClient?.initiateUser(
+          initiateUserRequest,
+        );
+
+        if (initiateUserResponse?.success) {
+          // nvaigation flow
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:46]: spelling mistake `navigation`

File: ChatSX/components/ParseDeepLink/index.ts [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+          let routes = getRoute(internalRoute);
```
Comment (ishaan1607) [inline ChatSX/components/ParseDeepLink/index.ts:47]: where is this `getRoute` written, it is not present in this PR.

File: ChatSX/screens/HomeFeed/index.tsx [hunk 22]
```diff
@@ -154,6 +158,31 @@ const HomeFeed = ({navigation}: Props) => {
     fetchData();
   }, [navigation, myClient]);
 
+  useEffect(() => {
+    const subscription = Linking.addEventListener('url', ({url}) => {
+
+      const UUID = users[0]?.userUniqueID;
```
Comment (ishaan1607) [inline ChatSX/screens/HomeFeed/index.tsx:164]: rename to `uuid`

File: ChatSX/screens/HomeFeed/index.tsx [hunk 23]
```diff
@@ -154,6 +158,31 @@ const HomeFeed = ({navigation}: Props) => {
     fetchData();
   }, [navigation, myClient]);
 
+  useEffect(() => {
+    const subscription = Linking.addEventListener('url', ({url}) => {
```
Comment (ishaan1607) [inline ChatSX/screens/HomeFeed/index.tsx:162]: why `subscription`, can we name the variable as `listener` as it is `Linking.addEventListener` is returning a `listener`

File: ChatSX/shareUtils/index.ts [hunk 24]
```diff
@@ -0,0 +1,35 @@
+import {Alert, Share} from 'react-native';
+import {ShareChatroomRequest} from '../Models/CommonFunctions';
+import {VALID_URI_REGEX} from '../constants/Regex';
+
+// this method generates URL for share
+export const shareChatroomURL = ({
+  chatroomId,
+  domain,
+}: ShareChatroomRequest) => {
+  let URL = `${domain}/chatroom?chatroom_id=${chatroomId}`;
```
Comment (ishaan1607) [inline ChatSX/shareUtils/index.ts:10]: rename to `url`, why it need to all caps?

**Review by ishaan1607 — CHANGES_REQUESTED** [review]

File: ChatSX/components/ParseDeepLink/index.ts (hunk 21 above)
Comment (jaixchaudhary) [reply ChatSX/components/ParseDeepLink/index.ts:48]: getRoute is the same that we use for the notification flow

**Review by ishaan1607 — CHANGES_REQUESTED** [review]

**Review by ishaan1607 — APPROVED** [review]

**Review by ishaan1607 — APPROVED** [review]

**Review by ishaan1607 — APPROVED** [review]

**Pull request merged**